com.vrcfury.vrcfury/Editor/VF/Feature/ParameterSmoothingBuilder.cs

```diff
@@ -0,0 +1,50 @@
+using System.Linq;
+using VF.Feature.Base;
+using VF.Injector;
+using VF.Service;
+using VF.Utils;
+using VF.Builder;
+using VRC.SDK3.Avatars.Components;
+using System.Collections.Concurrent;
+
+namespace VF.Feature
+{
+    public class ParameterSmoothingBuilder : FeatureBuilder
+    {
+
+        [VFAutowired] private readonly ParamSmoothingService smoothing;
+
+        [FeatureBuilderAction(FeatureOrder.ParameterSmoothing)]
+        public void Apply()
+        {
+            var avatar = avatarObject.GetComponent<VRCAvatarDescriptor>();
+
+            foreach (var c in manager.GetAllUsedControllers()) {
+                ReplaceUsagesOfSmoothedParameters(c);
+            }
+        }
+
+         private void ReplaceUsagesOfSmoothedParameters(ControllerManager c)
+        {
+            var smoothedParams = c.GetSmoothedParams().ToDictionary(v => v.name, v => v);
+            if (smoothedParams.Count == 0) {
+                return;
+            }
+            var raw = c.GetRaw();
+            var paramsMap = raw.parameters.ToDictionary(x => x.name, x => x);
+            var smoothedParamCache = new ConcurrentDictionary<string, string>();
+            raw.RewriteParameters(param => {
```

Comment:
Just found another issue here, when the smoothing service adds a new layer, it gets detected by the `RewriteParameters` call as well and rewrites the input of the smoothing blend tree as well.

I suppose there are two ways to fix this:

1. Make `RewriteParameters` collect all data first and then iterate over them, or
2. Make it so you can control the name of the smoothed parameter and add the smoothing layers after rewriting them